flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FunctionDefinitionFactory.java

```diff
@@ -34,4 +35,29 @@ public interface FunctionDefinitionFactory {
      * @return a {@link FunctionDefinition}
      */
     FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction);
+
+    /**
+     * Creates a {@link FunctionDefinition} from given {@link CatalogFunction}. If the {@link
+     * CatalogFunction} is created by user defined resource, the user of {@link
+     * FunctionDefinitionFactory} needs to override this method explicitly. The implementation logic
+     * needs to use the user classloader to load custom classes instead of the thread context
+     * classloader.
+     *
+     * @param name name of the {@link CatalogFunction}
+     * @param catalogFunction the catalog function
+     * @param userClassLoader the class loader is used to load user defined function's class
+     * @return a {@link FunctionDefinition}
+     */
+    default FunctionDefinition createFunctionDefinition(
+            String name, CatalogFunction catalogFunction, ClassLoader userClassLoader) {
+        if (!CollectionUtil.isNullOrEmpty(catalogFunction.getFunctionResources())) {
```

Comment:
In the case of user bundle his job with the custom function into a jar job , we also need to use the UserCodeClassLoader to load the class for these functions. 

So such check may be not enough. Can we not provide the default implementation?

Reply:
I think your point is a case, throw exception directly?